File: src/main/java/io/github/mg138/tsbook/listener/event/damage/DamageHandler.java
package io.github.mg138.tsbook.listener.event.damage;

import io.github.mg138.tsbook.Book;
import io.github.mg138.tsbook.entities.effect.EffectHandler;
import io.github.mg138.tsbook.entities.effect.data.StatusEffectType;
import io.github.mg138.tsbook.items.ItemIdentification;
import io.github.mg138.tsbook.items.ItemInstance;
import io.github.mg138.tsbook.items.ItemStats;
import io.github.mg138.tsbook.items.ItemUtils;
import io.github.mg138.tsbook.items.data.stat.StatMap;
import io.github.mg138.tsbook.items.data.stat.StatSingle;
import io.github.mg138.tsbook.items.data.stat.map.DamageDefenseRelation;
import io.github.mg138.tsbook.items.data.stat.set.*;
import io.github.mg138.tsbook.items.data.stat.StatType;
import io.github.mg138.tsbook.entities.util.MobType;
import io.github.mg138.tsbook.listener.event.damage.utils.CustomDamageEvent;
import io.github.mg138.tsbook.listener.event.damage.utils.DamageIndicator;
import io.github.mg138.tsbook.listener.event.damage.utils.StatCalculator;
import io.github.mg138.tsbook.players.ArcticPlayerDataService;
import io.github.mg138.tsbook.players.data.PlayerData;
import io.lumine.xikage.mythicmobs.MythicMobs;
import io.lumine.xikage.mythicmobs.api.bukkit.BukkitAPIHelper;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.attribute.Attribute;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.*;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataContainer;

import java.util.*;

public class DamageHandler {
    private static final BukkitAPIHelper mythicMobHelper = MythicMobs.inst().getAPIHelper();
    private static final Random rand = new Random();

    public static final Map<Player, Long> damageCD = new HashMap<>();

    public static void damagedByEntity(EntityDamageByEntityEvent event, Map<StatType, Double> defense) {
        Entity damaged = event.getEntity();
        if (!(damaged instanceof LivingEntity)) return;

        Entity damager = event.getDamager();
        if (mythicMobHelper.isMythicMob(damager)) {
            ConfigurationSection mob = Book.Companion.getCfg().getMMMob(mythicMobHelper.getMythicMobInstance(damager).getType().getInternalName());
            if (mob != null) {
                Map<StatType, StatMap> map = new HashMap<>();
                Map<StatType, Float> identification = new HashMap<>();

                for (String literalType : mob.getKeys(false)) {
                    StatType type = StatType.valueOf(literalType.toUpperCase());
                    map.put(type, new StatMap(type, new StatSingle(mob.getDouble(literalType))));
                    identification.put(type, 1F);
                }
                ItemStats[] stats = { new ItemStats(map, new ItemIdentification(identification), Book.Companion.getCfg()) };
                complexDamage(event, stats, defense);
            }
        } else if (damager instanceof Player) {
            Player player = (Player) damager;
            if (damageCD.putIfAbsent(player, System.currentTimeMillis()) != null) {
                if ((System.currentTimeMillis() - damageCD.get(player)) < 500) {
                    event.setCancelled(true);
                    return;
                }
                damageCD.replace(player, System.currentTimeMillis());
            }

            List<ItemStats> stats = new ArrayList<>();

            EntityEquipment equipment = player.getEquipment();
            if (equipment != null) {
                ItemStack item = equipment.getItemInMainHand();

                if (item.getType() != Material.AIR && item.getItemMeta() != null) {
                    ItemInstance inst = ItemUtils.getInstByItem(Book.inst, item);

                    if (inst != null) {
                        stats.add(inst.getStats());
                    }
                }
            }

            PlayerData data = ArcticPlayerDataService.dataServiceInstance.getData(player, ArcticPlayerDataService.Companion.getPlayerDataRef());
            if (data != null) {
                data.getEquipment().forEach((i, armor) -> stats.add(armor.getStats()));
            }

            complexDamage(event, stats.toArray(new ItemStats[0]), defense);
        } else if (damager instanceof Arrow) {
            Arrow arrow = (Arrow) damager;
            PersistentDataContainer container = arrow.getPersistentDataContainer();

            UUID[] uuids = container.get(new NamespacedKey(Book.inst, "item_uuids"), ItemUtils.UUID_ARRAY_TAG_TYPE);
            if (uuids == null) return;

            List<ItemStats> stats = new ArrayList<>();

            for (UUID uuid : uuids) {
                ItemInstance inst = ItemUtils.UUID_ITEM.get(uuid);
                if (inst == null) continue;

                stats.add(inst.getStats());
            }

            complexDamage(event, stats.toArray(new ItemStats[0]), defense);
        }
    }

    public static void complexDamage(EntityDamageByEntityEvent event, ItemStats[] stats, Map<StatType, Double> defense) {
        event.setDamage(0);
        LivingEntity livingEntity = (LivingEntity) event.getEntity();

        if (stats == null) {
            simpleDamage(livingEntity, 1);
            return;
        }

        double damage = 0.0;
        Entity entityDamager = event.getDamager();
        LivingEntity livingEntityDamager = null;

        CustomDamageEvent customDamageEvent = null;

        if (entityDamager instanceof LivingEntity) {
            livingEntityDamager = (LivingEntity) entityDamager;
            customDamageEvent = new CustomDamageEvent(livingEntity, livingEntityDamager);
        } else if (entityDamager instanceof Projectile) {
            if (((Projectile) entityDamager).getShooter() instanceof LivingEntity) {
                livingEntityDamager = (LivingEntity) ((Projectile) entityDamager).getShooter();
                customDamageEvent = new CustomDamageEvent(livingEntity, livingEntityDamager);
            }
        }

        double usedModifier = 0.0;
        for (Map.Entry<StatType, Double> entry : getModifier(stats).entrySet()) {
            StatType type = entry.getKey();
            Double modifier = entry.getValue();
            switch (type) {
                case MODIFIER_HELL:
                    if (MobType.isHellish(livingEntity.getType())) usedModifier += (modifier / 100);
                    break;
                case MODIFIER_MOBS:
                    if (MobType.isMob(livingEntity.getType())) usedModifier += (modifier / 100);
                    break;
                case MODIFIER_PLAYER:
                    if (livingEntity.getType() == EntityType.PLAYER) usedModifier += (modifier / 100);
                    break;
                case MODIFIER_ARTHROPOD:
                    if (MobType.isArthropod(livingEntity.getType())) usedModifier += (modifier / 100);
                    break;
                case MODIFIER_UNDERWATER:
                    if (MobType.isWatery(livingEntity.getType())) usedModifier += (modifier / 100);
                    break;
                case MODIFIER_UNDEAD:
                    if (MobType.isUndead(livingEntity.getType())) usedModifier += (modifier / 100);
                    break;
            }
        }

        Map<StatType, Double> damages = getDamage(stats);
        double critDamage = getStat(EnumSet.of(StatType.POWER_CRITICAL), stats).getOrDefault(StatType.POWER_CRITICAL, 0.0);
        double critChance = getStat(EnumSet.of(StatType.CHANCE_CRITICAL), stats).getOrDefault(StatType.CHANCE_CRITICAL, 0.0);

        for (StatType damageType : damages.keySet()) {
            double currentDamage;

            if (damageType == StatType.DAMAGE_TRUE) {
                currentDamage = StatCalculator.calculateTrueDamage(
                        damages.get(StatType.DAMAGE_TRUE),
                        defense.getOrDefault(StatType.DEFENSE_TRUE, 0.0),
                        1 + usedModifier
                );
            } else {
                int critical = 0;
                critChance /= 100;
                while (true) {
                    boolean criticalSuccess = (rand.nextFloat() < critChance);
                    if (!criticalSuccess) break;
                    critChance -= 1;
                    critical++;
                }
                currentDamage = StatCalculator.calculateDamage(
                        damages.get(damageType),
                        defense.getOrDefault(DamageDefenseRelation.relationship.get(damageType), 0.0),
                        1 + ((critical * (1 + (critDamage / 100))) - 1) + usedModifier
                );
            }
            damage += currentDamage;
            if (customDamageEvent != null) {
                customDamageEvent.addDamage(damageType, currentDamage);
            }

            double chance = 0.25;
            int x = 0;
            switch (damageType) {
                case DAMAGE_TEMPUS:
                case DAMAGE_AQUA:
                case DAMAGE_IGNIS:
                case DAMAGE_LUMEN:
                case DAMAGE_TERRA:
                case DAMAGE_UMBRA: {
                    chance = (0.25 + getStat(EnumSet.of(StatType.AFFINITY_ELEMENT), stats)
                            .getOrDefault(StatType.AFFINITY_ELEMENT, 0.0) / 100);
                }
            }

            while (true) {
                boolean success = (rand.nextFloat() < chance);
                if (!success) break;
                chance -= 1;
                x++;
            }

            if (x != 0) {
                switch (damageType) {
                    case DAMAGE_IGNIS: {
                        double power = x * currentDamage / 8;
                        if (power < 20) break;
                        EffectHandler.apply(StatusEffectType.BURNING, livingEntity, power, (int) (currentDamage / 6));
                        break;
                    }
                    case DAMAGE_PHYSICAL: {
                        double power = x * currentDamage / 12;
                        if (power < 20) break;
                        EffectHandler.apply(StatusEffectType.BLEEDING, livingEntity, power, (int) (currentDamage / 14));
                        break;
                    }
                    case DAMAGE_TEMPUS: {
                        double power = x * currentDamage / 8;
                        if (power < 20) break;
                        EffectHandler.apply(StatusEffectType.PARALYSIS, livingEntity, power, (int) (600 * (currentDamage / (currentDamage + 1000))));
                        break;
                    }
                }
            }
        }

        Map<StatType, Double> effectPower = getEffectPower(stats);
        for (Map.Entry<StatType, Double> entry : getEffectChance(stats).entrySet()) {
            StatType type = entry.getKey();
            Double chance = entry.getValue();

            int x = 0;
            chance /= 100;
            while (true) {
                boolean success = (rand.nextFloat() < chance);
                if (!success) break;
                chance -= 1;
                x++;
            }
            if (x == 0) continue;

            switch (type) {
                case CHANCE_DRAIN: {
                    Double power = effectPower.get(StatType.POWER_DRAIN);
                    if (power == null) break;
                    if (livingEntityDamager == null) break;
                    double result = livingEntityDamager.getHealth() + (damage * Math.min(1, (power / 100)));
                    double maxHealth = Objects.requireNonNull(livingEntityDamager.getAttribute(Attribute.GENERIC_MAX_HEALTH)).getBaseValue();

                    livingEntityDamager.setHealth(Math.min(result, maxHealth));
                    break;
                }
                case CHANCE_SLOWNESS: {
                    Double power = effectPower.get(StatType.POWER_SLOWNESS);
                    if (power == null) break;
                    EffectHandler.apply(StatusEffectType.SLOWNESS, livingEntity, x * power / 100, (int) (x * power * 90));
                    break;
                }
                case CHANCE_LEVITATION: {
                    Double power = effectPower.get(StatType.POWER_LEVITATION);
                    if (power == null) break;
                    EffectHandler.apply(StatusEffectType.LEVITATION, livingEntity, 0, (int) (x * power * 20));
                    break;
                }
                case CHANCE_NAUSEOUS: {
                    Double power = effectPower.get(StatType.POWER_NAUSEOUS);
                    if (power == null) break;
                    EffectHandler.apply(StatusEffectType.NAUSEOUS, livingEntity, 0, (int) (x * power * 20));
                    break;
                }
            }
        }
        double health = livingEntity.getHealth();
        double result = health - damage;
        livingEntity.setHealth(result < 0 ? 0 : result);

        livingEntity.setMaximumNoDamageTicks(0);
        livingEntity.setNoDamageTicks(0);

        if (customDamageEvent != null) {
            Bukkit.getPluginManager().callEvent(customDamageEvent);
        }
    }

    private static Map<StatType, Double> getStat(Set<StatType> template, ItemStats[] stats) {
        Map<StatType, Double> result = new HashMap<>();
        for (ItemStats stat : stats) {
            template.forEach(type -> {
                if (stat.getStats().containsKey(type)) {
                    result.put(type, stat.getStat(type) + result.getOrDefault(type, 0.0));
                }
            });
        }
        return result;
    }

    public static Map<StatType, Double> getDefense(ItemStats[] stats) {
        return getStat(DefenseType.DEFENSE, stats);
    }

    public static Map<StatType, Double> getModifier(ItemStats[] stats) {
        return getStat(ModifierType.MODIFIER, stats);
    }

    public static Map<StatType, Double> getDamage(ItemStats[] stats) {
        return getStat(DamageType.DAMAGE, stats);
    }

    public static Map<StatType, Double> getEffectPower(ItemStats[] stats) {
        return getStat(EffectPowerType.EFFECT_POWER, stats);
    }

    public static Map<StatType, Double> getEffectChance(ItemStats[] stats) {
        return getStat(EffectChanceType.EFFECT_CHANCE, stats);
    }

    public static boolean simpleDamage(LivingEntity entity, double damage, StatType damageType, boolean display) {
        if (entity.isDead()) return false;

        double health = entity.getHealth();
        double result = health - damage;
        entity.setHealth(result < 0 ? 0 : result);
        entity.damage(0);

        entity.setMaximumNoDamageTicks(0);
        entity.setNoDamageTicks(0);

        if (display && DamageType.DAMAGE.contains(damageType)) {
            DamageIndicator.displayDamage(damage, damageType, entity.getLocation());
        }

        return true;
    }

    public static boolean simpleDamage(LivingEntity entity, double damage) {
        return simpleDamage(entity, damage, null, false);
    }
}

File: src/main/java/io/github/mg138/tsbook/listener/event/damage/DamageHandler.kt
package io.github.mg138.tsbook.listener.event.damage

import io.github.mg138.tsbook.Book
import io.github.mg138.tsbook.entities.effect.EffectHandler
import io.github.mg138.tsbook.entities.effect.data.StatusEffectType
import io.github.mg138.tsbook.utils.MobType
import io.github.mg138.tsbook.items.ItemStats
import io.github.mg138.tsbook.items.ItemUtils
import io.github.mg138.tsbook.stat.*
import io.github.mg138.tsbook.stat.util.StatUtil
import io.github.mg138.tsbook.stat.util.set.*
import io.github.mg138.tsbook.listener.event.damage.utils.CustomDamageEvent
import io.github.mg138.tsbook.listener.event.damage.utils.DamageIndicator
import io.github.mg138.tsbook.players.ArcticGlobalDataService
import io.github.mg138.tsbook.players.data.PlayerData
import io.github.mg138.tsbook.setting.mob.MobConfig
import io.github.mg138.tsbook.stat.util.map.DamageDefenseRelation
import io.lumine.xikage.mythicmobs.MythicMobs
import io.lumine.xikage.mythicmobs.api.bukkit.BukkitAPIHelper
import org.bukkit.Bukkit
import org.bukkit.attribute.Attribute
import org.bukkit.entity.*
import org.bukkit.event.entity.EntityDamageByEntityEvent
import java.util.*
import kotlin.collections.HashMap
import kotlin.math.min

//TODO make it prettier :/
object DamageHandler {
    val mythicMobHelper: BukkitAPIHelper = MythicMobs.inst().apiHelper
    private val rand = Random()
    private val damageCD: MutableMap<Player, Long> = HashMap()

    fun remove(player: Player) {
        damageCD.remove(player)
    }

    fun unload() {
        damageCD.clear()
    }

    fun damagedByEntity(event: EntityDamageByEntityEvent, defense: StatMap) {
        val damager = event.damager

        when {
            mythicMobHelper.isMythicMob(damager) -> {
                MobConfig[mythicMobHelper.getMythicMobInstance(damager).type.internalName]?.let {
                    complexDamage(event, it.itemStats.stats, defense)
                }
            }
            damager is Player -> {
                if (System.currentTimeMillis() - damageCD.getOrDefault(damager, 0L) <= 500) {
                    event.isCancelled = true
                    return
                }
                damageCD[damager] = System.currentTimeMillis()

                val itemStats: MutableList<ItemStats> = LinkedList()
                damager.equipment?.itemInMainHand?.let { item ->
                    if (ItemUtils.checkItem(item)) {
                        ItemUtils.getInstByItem(Book.inst, item)?.itemStats?.let { itemStats.add(it) }
                    }
                }
                ArcticGlobalDataService.inst.getData<PlayerData>(damager, PlayerData::class)
                    ?.equipment?.forEach { _, armor -> armor.itemStats?.let { itemStats.add(it) } }

                complexDamage(event, StatUtil.combine(itemStats), defense)
            }
            damager is Arrow -> {
                val itemStats: MutableList<ItemStats> = LinkedList()
                damager.persistentDataContainer[ItemUtils.uuidArrayKey, ItemUtils.uuidArrayTag]
                    ?.forEach { uuid ->
                        ItemUtils.itemCache[uuid]?.let { inst ->
                            inst.itemStats?.let { itemStats.add(it) }
                        }
                    }
                complexDamage(event, StatUtil.combine(itemStats), defense)
            }
        }
    }

    fun complexDamage(event: EntityDamageByEntityEvent, stats: StatMap, defense: StatMap) {
        val entity = event.entity as LivingEntity

        val damager = when (val entityDamager = event.damager) {
            is LivingEntity -> entityDamager
            is Projectile -> {
                val shooter = entityDamager.shooter
                if (shooter is LivingEntity) shooter else return
            }
            else -> return
        }

        val customEvent = CustomDamageEvent(entity, damager)

        var usedModifier = 0.0
        StatUtil.getModifier(stats).forEach { (type, modifier) ->
            when (type) {
                StatType.MODIFIER_HELL -> if (MobType.isHellish(entity.type)) usedModifier += modifier.stat / 100
                StatType.MODIFIER_MOBS -> if (MobType.isMob(entity.type)) usedModifier += modifier.stat / 100
                StatType.MODIFIER_PLAYER -> if (entity.type == EntityType.PLAYER) usedModifier += modifier.stat / 100
                StatType.MODIFIER_ARTHROPOD -> if (MobType.isArthropod(entity.type)) usedModifier += modifier.stat / 100
                StatType.MODIFIER_UNDERWATER -> if (MobType.isWatery(entity.type)) usedModifier += modifier.stat / 100
                StatType.MODIFIER_UNDEAD -> if (MobType.isUndead(entity.type)) usedModifier += modifier.stat / 100
                else -> Unit
            }
        }

        val damageSum = damage(stats, defense, usedModifier, customEvent)
        elementEffect(stats, entity, usedModifier)
        effect(stats, entity, damager, damageSum)

        event.damage = damageSum
        entity.maximumNoDamageTicks = 0
        entity.noDamageTicks = 0
        Bukkit.getPluginManager().callEvent(customEvent)
    }

    fun simpleDamage(
        entity: LivingEntity,
        damage: Double,
        damageType: StatType = StatType.DAMAGE_NONE,
        display: Boolean = false
    ): Boolean {
        if (entity.isDead) return false
        if (damage == 0.0) return true
        entity.damage(damage)
        entity.maximumNoDamageTicks = 0
        entity.noDamageTicks = 0
        if (display && DamageType.types.contains(damageType)) {
            DamageIndicator.displayDamage(damage, damageType, entity.location)
        }
        return true
    }

    private fun effect(stats: StatMap, entity: LivingEntity, damager: LivingEntity, damageSum: Double) {
        val effectChance = StatUtil.getEffectChance(stats)
        if (effectChance.isEmpty()) return

        val effectPower = StatUtil.getEffectPower(stats)

        effectChance.forEach { (type, rawChance) ->
            val strikes = StatUtil.intCheckBelowZero(
                stat = rawChance.stat,
                above = {
                    val temp = it / 100
                    temp.toInt() + if (rand.nextDouble() < temp % 1) 1 else 0
                }
            )

            if (strikes > 0) {
                when (type) {
                    StatType.CHANCE_DRAIN -> {
                        val rawPower = effectPower.getStatOut(StatType.POWER_DRAIN)
                        if (rawPower > 0) {
                            val result = damager.health + damageSum * min(rawPower / 100, 1.0)
                            val maxHealth = damager.getAttribute(Attribute.GENERIC_MAX_HEALTH)!!.baseValue
                            damager.health = min(result, maxHealth)
                        }
                    }
                    StatType.CHANCE_SLOWNESS -> {
                        val rawPower = effectPower.getStatOut(StatType.POWER_SLOWNESS)
                        EffectHandler.apply(
                            StatusEffectType.SLOWNESS,
                            entity,
                            strikes * rawPower / 100,
                            (strikes * rawPower * 90).toLong()
                        )
                    }
                    StatType.CHANCE_LEVITATION -> {
                        val ticks = 20 * strikes * effectPower.getStatOut(StatType.POWER_LEVITATION)
                        if (ticks >= 20) EffectHandler.apply(StatusEffectType.LEVITATION, entity, 0.0, ticks.toLong())
                    }
                    StatType.CHANCE_NAUSEOUS -> {
                        val ticks = 20 * strikes * effectPower.getStatOut(StatType.POWER_NAUSEOUS)
                        if (ticks >= 20) EffectHandler.apply(StatusEffectType.NAUSEOUS, entity, 0.0, ticks.toLong())
                    }
                    else -> Unit
                }
            }
        }
    }

    private fun damage(stats: StatMap, defense: StatMap, modifier: Double, event: CustomDamageEvent? = null): Double {
        val damages = StatUtil.getDamage(stats)
        if (damages.isEmpty()) return 0.0

        var certainCritStrikes = 0
        val critDamage = StatUtil.doubleCheckBelowZero(
            stat = stats.getStatOut(StatType.POWER_CRITICAL),
            above = { it / 100 }
        )
        val critChance = StatUtil.doubleCheckBelowZero(
            stat = stats.getStatOut(StatType.CHANCE_CRITICAL),
            above = {
                val temp = it / 100
                certainCritStrikes = temp.toInt()
                temp % 1
            }
        )

        println(critChance)
        println(critDamage)

        var damageSum = 0.0
        damages.forEach { (damageType, rawDamage) ->
            val critStrikes = certainCritStrikes + (if (rand.nextDouble() < critChance) 1 else 0)
            val damage = when (damageType) {
                StatType.DAMAGE_TRUE -> {
                    StatUtil.calculateTrueDamage(
                        damage = rawDamage.stat,
                        defense = defense.getStatOut(StatType.DEFENSE_TRUE),
                        modifier = 1 + modifier
                    )
                }
                else -> {
                    StatUtil.calculateDamage(
                        damage = rawDamage.stat,
                        defense = defense.getStatOut(DamageDefenseRelation.relation[damageType]!!),
                        modifier = 1 + modifier + (critDamage * critStrikes)
                    )
                }
            }
            event?.addDamage(damageType, damage)
            damageSum += damage
        }
        return damageSum
    }

    private fun elementEffect(stats: StatMap, entity: LivingEntity, modifier: Double) {
        val elementDamages = StatUtil.getElementDamage(stats)
        if (elementDamages.isEmpty()) return

        var certainStrike = 0
        val chance = StatUtil.doubleCheckBelowZero(
            stat = (0.25 + stats.getStatOut(StatType.AFFINITY_ELEMENT)),
            above = {
                val temp = it / 100
                certainStrike = temp.toInt()
                temp % 1
            }
        )

        elementDamages.forEach { (damageType, rawDamage) ->
            val damage = StatUtil.calculateModifier(rawDamage.stat, modifier)
            val strikes = (certainStrike + if (rand.nextDouble() < chance) 1 else 0)
            when (damageType) {
                StatType.DAMAGE_IGNIS -> {
                    val power = (damage / 8) * strikes
                    val tick = (damage / 6).toLong()
                    if (power > 20 && tick > 10) {
                        EffectHandler.apply(StatusEffectType.BURNING, entity, power, tick)
                    }
                }
                StatType.DAMAGE_PHYSICAL -> {
                    val power = (damage / 12) * strikes
                    val tick = (damage / 14).toLong()
                    if (power > 20 && tick > 10) {
                        EffectHandler.apply(StatusEffectType.BLEEDING, entity, power, tick)
                    }
                }
                StatType.DAMAGE_TEMPUS -> {
                    val power = (damage / 8) * strikes
                    val tick = (600 * (damage / (damage + 1000))).toLong()
                    if (power > 20 && tick > 60) {
                        EffectHandler.apply(StatusEffectType.PARALYSIS, entity, power, tick)
                    }
                }
                else -> Unit
            }
        }
    }
}